crypto/key_manager.go

```diff
@@ -82,140 +59,84 @@ func (k PEMKeyManager) HashAlgorithm() crypto.Hash {
 	return crypto.SHA256
 }
 
-// LoadPrivateKey loads a private key from a PEM encoded string, decrypting it if necessary
-func (k *PEMKeyManager) LoadPrivateKey(pemEncodedKey, password string) error {
-	block, rest := pem.Decode([]byte(pemEncodedKey))
-	if len(rest) > 0 {
-		return errors.New("extra data found after PEM decoding")
-	}
-
-	der := block.Bytes
-	if password != "" {
-		pwdDer, err := x509.DecryptPEMBlock(block, []byte(password))
-		if err != nil {
-			return err
-		}
-		der = pwdDer
-	}
-
-	key, algo, err := parsePrivateKey(der)
-	if err != nil {
-		return err
-	}
-
-	k.serverPrivateKey = key
-	k.signatureAlgorithm = algo
-	signer, err := k.Signer()
-	if err != nil {
-		return err
-	}
-	k.serverPublicKey = signer.Public()
-	return nil
+// Signer returns a signer based on our private key.
+func (k PEMKeyManager) Signer() crypto.Signer {
+	return k.signer
 }
 
-// LoadPublicKey loads a public key from a PEM encoded string.
-func (k *PEMKeyManager) LoadPublicKey(pemEncodedKey string) error {
-	publicBlock, rest := pem.Decode([]byte(pemEncodedKey))
-	if publicBlock == nil {
-		return errors.New("could not decode PEM for public key")
-	}
-	if len(rest) > 0 {
-		return errors.New("extra data found after PEM key decoded")
-	}
-
-	k.rawPublicKey = publicBlock.Bytes
-
-	parsedKey, err := x509.ParsePKIXPublicKey(publicBlock.Bytes)
-	if err != nil {
-		return errors.New("unable to parse public key")
-	}
+// PublicKey returns the public key corresponding to the private key.
+func (k PEMKeyManager) PublicKey() crypto.PublicKey {
+	return k.signer.Public()
+}
 
-	k.serverPublicKey = parsedKey
-	return nil
+// NewFromPrivateKey creates PrivateKeyManager using a private key.
+func NewFromPrivateKey(key crypto.PrivateKey) (PrivateKeyManager, error) {
+	var signer crypto.Signer
+	var sigAlgo sigpb.DigitallySigned_SignatureAlgorithm
+
+	switch key.(type) {
+	case *ecdsa.PrivateKey:
+		signer = key.(crypto.Signer)
+		sigAlgo = sigpb.DigitallySigned_ECDSA
+	case *rsa.PrivateKey:
+		signer = key.(crypto.Signer)
+		sigAlgo = sigpb.DigitallySigned_RSA
+	default:
+		return nil, errors.New("unsupported key type")
+	}
+
+	return &PEMKeyManager{
```

Comment:
No, I'm fine with leaving this as if. Any comment that starts "could" or "might" is optional. I think fewer variables are also better for readability and maintenance, but it really doesn't make much difference here. I'll just point out that almost all of the examples of switches in Effective Go use my proposed pattern though: https://golang.org/doc/effective_go.html#switch